examples/triplet/convert_3d_triplet_data.cpp

```diff
@@ -0,0 +1,163 @@
+  leveldb::Status status = leveldb::DB::Open(
+      options, db_filename, &db);
+  CHECK(status.ok()) << "Failed to open leveldb " << db_filename
+      << ". Is it already existing?";
+
+  char* label_temp = new char[4];  // label for unsigned char*
+  signed char* label_i = new signed char[4];  // label for triplet
+  signed char* label_j = new signed char[4];
+  signed char* label_k = new signed char[4];
+  signed char* label_l = new signed char[4];  // label for pair wise
+  signed char* label_m = new signed char[4];
+  char* pixels = new char[5 * rows * cols];
+  const int kMaxKeyLength = 10;
+  char key[kMaxKeyLength];
+  std::string value;
+
+  caffe::Datum datum;
+  datum.set_channels(5);  // one channel for each image in the triplet and pair
+  datum.set_height(rows);
+  datum.set_width(cols);
+  LOG(INFO) << "A total of " << num_items << " items.";
+  LOG(INFO) << "Rows: " << rows << " Cols: " << cols;
+  for (unsigned int itemid = 0; itemid < 10 * num_items; ++itemid) {\
+    int i = caffe::caffe_rng_rand() % num_items;  // pick triplet groups
+    int j = caffe::caffe_rng_rand() % num_items;
+    int k = caffe::caffe_rng_rand() % num_items;
+    int l = caffe::caffe_rng_rand() % num_items;  // pick pair wise groups
+    int m = caffe::caffe_rng_rand() % num_items;
+    read_image(&image_file, &label_file, i, rows, cols,  // read triplet
+        pixels, label_temp, label_i);
+    read_image(&image_file, &label_file, j, rows, cols,
+        pixels + (rows * cols), label_temp, label_j);
+    read_image(&image_file, &label_file, k, rows, cols,
+        pixels + (2 * rows * cols), label_temp, label_k);
+    read_image(&image_file, &label_file, l, rows, cols,  // read pair wise
+        pixels + (3 * rows * cols), label_temp, label_l);
+    read_image(&image_file, &label_file, m, rows, cols,
+        pixels + (4 * rows * cols), label_temp, label_m);
+
+    datum.set_data(pixels, 5*rows*cols);  // set data
+    bool triplet_class_pass = false;
+    bool triplet_class_same = false;
+    bool triplet_pose_pass = false;
+    bool pair_class_pass = false;
+
+    int ij_diff_x = static_cast<int>(*(label_i+1)-*(label_j+1));
+    int ij_diff_y = static_cast<int>(*(label_i+2)-*(label_j+2));
+    int ij_diff_z = static_cast<int>(*(label_i+3)-*(label_j+3));
+    int ik_diff_x = static_cast<int>(*(label_i+1)-*(label_k+1));
+    int ik_diff_y = static_cast<int>(*(label_i+2)-*(label_k+2));
+    int ik_diff_z = static_cast<int>(*(label_i+3)-*(label_k+3));
+
+    int ij_x = ij_diff_x*ij_diff_x;
+    int ij_y = ij_diff_y*ij_diff_y;
+    int ij_z = ij_diff_z*ij_diff_z;
+    int ik_x = ik_diff_x*ik_diff_x;
+    int ik_y = ik_diff_y*ik_diff_y;
+    int ik_z = ik_diff_z*ik_diff_z;
+
+    int dist_ij = ij_x + ij_y + ij_z;
+    int dist_ik = ik_x + ik_y + ik_z;
+    if ((*label_i  == *label_j) && (*label_i  == *label_k))
+      triplet_class_same = true;
+    if ((dist_ij < dist_ik) && (triplet_class_same))
+      triplet_pose_pass = true;
```

Comment:
I modified the codes and now it will use different distance as threshold to verify the similarity. As my sphere all have a radius of 100, I set the similar distance as less than 30 degree in view point which means a nearly 100 * sin(pi/60) = 50 distance on sphere and the dissimilar distance as more than 90 degree in view point which means a 100 * sqrt(2) = 141 on sphere.
I haven't have a count on the edges as you have described because the edges is not enough in a simple sphere and the edges are also not perfect as expected because of the method to eliminating duplex points generated in different iterations also eliminated some useful points, maybe it will be solved later.